### PR TITLE
Remove attacher stateful set

### PR DESCRIPTION
**Pull request by Madhu-1**
Currently, we deploy attacher stateful set as part of our deployment, we no longer need it, added cluster driver registrar container to the provisioner stateful set to register csidriver CRD.

more info in #278 
Thanks, @ShyamsundarR for bringing up this issue.

Fixes: #278 

@rootfs @kfox1111 @gman0 @ShyamsundarR @humblec @j-griffith  PTAL

### CODE REVIEW

**Comment by rootfs** [comment]
would `provisioner` still be a good name for all the controllers (provisioner, snapshotter, and attacher)?

**Comment by gman0** [comment]
Are `create` and `delete` verbs not enough for `apiextensions.k8s.io/customresourcedefinitions`?

**Comment by Madhu-1** [comment]
> still be a good name for all the controllers

@rootfs any suggestion?

**Comment by Madhu-1** [comment]
@gman0 I have tested with current template, everything works fine, need to remove get and list from CRD and test so added DNM label.

**Comment by rootfs** [comment]
@Madhu-1 how about `controllers`?

**Comment by kfox1111** [comment]
lets have a wider discussion...

(Not ceph-csi specific)
The trend of loading controllers into k8s and giving them enough access to self register crd's and other things has always felt a little bad. If your deploying with helm, helm is already dealing with registering objects and has all the permissions needed to register them. The permissions to do anything stop after its installed, so it can't be used to do further things to the cluster. This seems good. But, if you give a long running process permissions to do such things, then it can be used to do bad things later. Another facet is auditing. If its in helm, it can be easily parsed by a human and verified its safe before running. This is very hard to do if its buried inside code.

(ceph-csi specific)
That sidecar is marked 1.14 only. Thats dropping support for 1.13 very quickly...

I think this code should be made conditional in the helm chart. If 1.13, use the attacher. if 1.14+, then register the csidriver object directly. No extra permissions needed.

**Comment by ShyamsundarR** [comment]
> (ceph-csi specific)
> That sidecar is marked 1.14 only. Thats dropping support for 1.13 very quickly...
> 
> I think this code should be made conditional in the helm chart. If 1.13, use the attacher. if 1.14+, then register the csidriver object directly. No extra permissions needed.

After reading through the need for an [attacher](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/container-storage-interface.md#attaching-and-detaching) even in the absence of the CSI plugin supporting Controller[Publish/Unpublish]Volume RPC and capability, I think we should retain the attacher for v1.13 deployments, as mentioned by @kfox1111.

**Comment by humblec** [comment]
I do second  @kfox1111's thought . That said, IMO, the cluster-driver-registrar may be good for future enhancement of csi with cluster CRDs and it may be the way to go in upcoming releases. However at this stage, the cluster driver registrar and cluster wide operation looks immature to me and relying on that wouldnt a good idea. My 2 cents :)

**Comment by kfox1111** [comment]
I raised an issue here to track it at the cluster-driver-registrar level:
https://github.com/kubernetes-csi/cluster-driver-registrar/issues/44

**Comment by Madhu-1** [comment]
sent a PR to move attacher container to provisioner pod, will revisit this in later.

**Comment by sbskas** [comment]
I tested the patch and had to update the csi-provisionner. It's missing the authorisation to update persistentvolumes.

**Comment by Madhu-1** [comment]
> I tested the patch and had to update the csi-provisionner. It's missing the authorisation to update persistentvolumes.

@sbskas This PR doesn't remove any of the roles for the provisioner if something is missing please create an issue for it.

File: deploy/cephfs/helm/templates/provisioner-clusterrole.yaml [hunk 1]
```diff
@@ -31,4 +31,10 @@ rules:
   - apiGroups: ["csi.storage.k8s.io"]
     resources: ["csinodeinfos"]
     verbs: ["get", "list", "watch"]
+  - apiGroups: ["csi.storage.k8s.io"]
+    resources: ["csidrivers"]
```
Comment (kfox1111) [inline deploy/cephfs/helm/templates/provisioner-clusterrole.yaml]: Can be removed I think?

**Pull request closed without merging**